Implement neptune-autopr

mathias-baumann-sociomantic opened the pull request.

mathias-baumann-sociomantic commented:
Forgot some test code that is now removed.
CI seems to fail because of the d-yaml update

gautam-kotian-sociomantic commented on src/autopr/SubModUpdate.d:
```diff
@@ -0,0 +1,182 @@
+updates submodule **%s** from version **%s** to version **%s**
+
+You can modify this behavior using your `.neptune.yml` file.
+See the according [documentation](https://github.com/sociomantic-tsunami/neptune/blob/master/doc/neptune-metadata.rst#automatic-update-prs).
```
`See the according documentation` sounds weird. It might be better to keep it simple as:
```
You can modify this behavior using your `.neptune.yml` file as described [here](https://github.com/sociomantic-tsunami/neptune/blob/master/doc/neptune-metadata.rst#automatic-update-prs).
```

mathias-baumann-sociomantic commented:
Updated:
- Rebased on latest master
- modified commit message according to https://github.com/sociomantic-tsunami/neptune/issues/159
- rephrased PR message according to gautams suggestion
- added check to prevent redundant forking

mathias-baumann-sociomantic commented:
I can haz review?

joseph-wakeling-sociomantic commented:
3 key things I'd like to see:

* author/committer info clearly indicating that these commits were generated by `neptune-autopr`, so that responsibility is clearly defined (I don't know if this depends on e.g. the source repo where the PR branch is defined, if so I presume that this will be resolved automatically)

* support for D2-only projects (i.e. projects whose `.D2-ready` file contains `ONLY`), so that we don't get updates from `vX.y.z+d2` release tags to `vX.y.z+1` D1 release tags

* separate PRs for stable releases and pre-releases, so the two are not mixed up

I don't mind if these things are implemented in follow-up PRs rather than here; I can open separate issues if so.

joseph-wakeling-sociomantic commented on src/autopr/SubModsUpdate.d:
```diff
@@ -0,0 +1,214 @@
+        writefln("%s> Created updated tree for %s (%s)",
+            this.repo, update.repo.name, tree_sha);
+
+        auto commit_msg = format("Advance %s from %s to %s\n\n%s %s(%s)...%s(%s)",
```
VERY bikesheddy, but I'd rather have `Update` than `Advance`.  But that's just because that matches the style that some of my projects are already using ;-)

mathias-baumann-sociomantic replied:
you'll need to fight with @ben-palmer-sociomantic over this one

joseph-wakeling-sociomantic replied:
Semantically, I think `Update` is unarguable fact, whereas `Advance` is a matter of opinion ;-)  But it's VERY bikesheddy, so feel free to ignore.

mathias-baumann-sociomantic commented:
> author/committer info clearly indicating that these commits were generated by neptune-autopr

During the testing phase I am using my account. When things seem stable it will be done by a dedicated github account which will act as author of both PR and commits

> support for D2-only

Noted and issue created :>

> separate PRs

is currently being discussed

joseph-wakeling-sociomantic commented:
> When things seem stable it will be done by a dedicated github account which will act as author of both PR and commits

OK, works for me, and obviously keeps the tool easy to use for others outside Sociomantic.

mathias-baumann-sociomantic commented:
Update: Fixed a minor bug about detection of existing PRs

mathias-baumann-sociomantic commented:
I would appreciate if someone would merge this at some point

mathias-baumann-sociomantic commented:
Yay
